### DISCUSSION TITLE
Calling a job from another service

### DISCUSSION THREAD

**Discussion by CrochetFeve0251** [lucidarch/lucid#34] [Q&A] [answered]
I am trying to write an application with micro service using lucid. However I have quite a problem getting which code I need to put in the job from a Service that call a job from another service.

Do I need to repeat the same code or is there something to write inside it that send it directly to the queue?

Sorry by advance if the question seems stupid but I didn't found my answer by reading the documentation.

Edit: I found a video explaining Lucid and I found my way to see services was wrong. However I have now another question in my head: Are all services on the same database? If it is the case how to handle migrations?

**Comment by Mulkave** [comment, marked as answer]
First of all, no question is stupid. Please ask anything that comes to mind 🙂

**Preface**
To get to your question, please keep in mind that when building microservices, in most cases a single service will be an application carrying a single responsibility. Services in Lucid are preferably used with monolith application but you can surely still work with them in a microservice.

**Use Case**
Before deciding to call a job from different services, there are some questions to ask your current structure to determine whether it is a valid case in the first place. Answering these questions would give you a better indication on the direction you should take. Also remember that there is no one-size fits all solution, it is always a matter of preference and Lucid is flexible enough to stay out of your way and let you do what you feel is best.

The questions are:

1. Should the functionality be within a job? Or can it be divided into several jobs and fit within an Operation? (since operations can cross domains but Services cannot)
2. Does this service carry enough functionality to be divided into its own service? or can we merge the two into one bigger service with a broader responsibility. (use this only when the responsibilities are extremely similar).
  ***Example**: Having a `Finance` and an `Inventory` service makes total sense, there's nothing that brings them together. However, having `Posts` and `Users` services aren't adequate, because they belong in Domains, and would rather have a `Publishing` service that takes care of both functionalities.*
3. Previous point leads to this question: Are my services valid as services? Or should they be transformed into Domains instead? To answer this question, we should go back to the very definitions of both Services and Domains to support our decision. The following is from the current version of the docs:

    **Services**
    > Services are structured for scale, by adding an encapsulation layer for features, operations, routes, controllers, commands, resources (views) providers, tests and database [...]

    > Their presence is necessary in multi-purpose projects that are set to scale in several areas of functionalities, a.k.a horizonal growth. They communicate with domains to compose jobs into groups of functionalities in their own features and operations.

    more here: https://docs.lucidarch.dev/services/

    **Domains**
    > Even though they differ in the nature of functionalities they expose, these are still domains from which our application would (re)use code wherever possible. This separation is only logical to help with the understanding of domains but is not an actual separation in code.

    more here: https://docs.lucidarch.dev/domains/


I am not sure whether this will be enough to answer your question, so to get more concrete with the answer I would appreciate more details, if possible.

**Reply by CrochetFeve0251** [reply]
Your answer correspond totally what I realized when I watched this video: https://www.youtube.com/watch?v=wSnM4JkyxPw (which has been done by you): I had a big confusion between domains and services when I asked the question.
Thanks a lot for the video and the answer it helped me to find real services from my application.

**Reply by CrochetFeve0251** [reply]
@Mulkave Just a last point, do you know any project that has been done with lucid and microservices that is open source (in public on github or gitlab) so I can refer to it when I got a problem?

**Reply by Mulkave** [reply]
Glad it helped 😄 
There is this project that's based on [the docs](https://docs.lucidarch.dev/getting-started/micro/) on [this repo](https://github.com/lucidarch/getting-started-micro) and @Batisska open sourced his [Library](https://github.com/Batisska/library) project. If anything comes up i'll add it here too. Meanwhile, please feel free to ask as many questions as you wish in the discussion board, it's what it's made for 😉
